Synthetic Data Creation - [WIP] (#60)
johko/computer-vision-course: course/4 Computer Vision in the Wild/4.2 Synthetic Data Creation/README.md

Synthetic data is artificially generated on a computer to augment or replace real data to improve AI models, protect sensitive data, and mitigate bias. Ideally, the synthetic data should have the same statistical properties as the real-world data it is supplementing.

## Why Would You Use Synthetic Data?

- Ethics/privacy issues(faces)/regulatory constraints
- Lack of availability of data (medical images, less common traffic signs, emergency vehicles, endangered animals)
- Expense/practicality of annotating real life data/time constraints

## Approaches to Generating Synthetic Data

- CAD & Blender : allows to create photorealistic image datasets of 3d scenes while controlling parameters. Allows to compute metrics by comparing to the ground truth (generation parameters). Very robust method but limited in generation quality/diversity/quantity

- Use cases include using [commercial applications](https://amazon-berkeley-objects.s3.amazonaws.com/static_html/ABO_CVPR2022.pdf), generating [synthetic faces](https://arxiv.org/abs/2109.15102), and [monitoring wildlife](https://openaccess.thecvf.com/content_CVPR_2020/papers/Mu_Learning_From_Synthetic_Animals_CVPR_2020_paper.pdf).

> - Deep generative models (Transformers/GANs/Diffusion models): allows to expend a dataset, balance it, solve privacy issues... Very conveniant and powerfull but can create datasets with biases, incoherence and repetitiveness which induces an important overtraining risk and produce a restricted set of predictions

Reviewer: ^ the answer to my first question becomes a bit clearer here

Author: What I would ideally like to demo for CAD / 3d is a random room generator or animal image generator.  While BlenderProc has a random room generator, and this should be simple, it loads a dataset no longer publicly available and the code is specific to that dataset and has to be changed.  I have spent a lot of time on this  but I am  taking a break for now.  

I really like the idea of generating animal images to train a model to monitor endangered species.  My challenge with implementing this demo is getting good textures for the animal. This is not my area of expertise, but my instinct is that  getting the textures right wouldn't be important if you were monitoring the populations of very different species, but probably would be if you wanted to monitor closely related species.  
Any thoughts?

> - Deep generative models (Transformers/GANs/Diffusion models): allows to expend a dataset, balance it, solve privacy issues... Very conveniant and powerfull but can create datasets with biases, incoherence and repetitiveness which induces an important overtraining risk and produce a restricted set of predictions

Reviewer 2: ```suggestion
- Deep generative models (Transformers/GANs/Diffusion models): allows expanding a dataset, tackle data imbalance, and solve privacy issues. Very convenient and powerful but can create datasets with biases, incoherence, and repetitiveness, which induces an important overtraining risk and produces a restricted set of predictions
```

The sentences are too long

### Possible notebook ideas

3D objects

[ShapeNet](https://huggingface.co/datasets/ShapeNet/ShapeNetCore)

Synthetic Faces

[Blog](https://huggingface.co/blog/train-your-controlnet)
[FaceSynthetics](https://github.com/microsoft/FaceSynthetics)
[face_synthetics_spiga dataset](https://huggingface.co/datasets/pcuenq/face_synthetics_spiga)
[SPIGA repo](https://github.com/andresprados/SPIGA#spiga-shape-preserving-facial-landmarks-with-graph-attention-networks)

Synthetic Animals

[Learning from Synthetic Animals](https://github.com/JitengMu/Learning-from-Synthetic-Animals)

- Motivation: monitoring wild animal behaviours, developing biologically inspired robots, motion capture systems
- Approach: Use synthetic CAD model and generate pseudo-labels on real images

[Demo notebook](https://github.com/JitengMu/Learning-from-Synthetic-Animals/blob/master/demo.ipynb)

> ### Medical Imaging

Reviewer: Note for later: could be cool to get Katie Link's opinion on this section once it fills out a bit more, even just ~5m

Motivation: Labeling is very expensive in medical imaging but rare cases are very important (to study rare illnesses) -> Synthetic data enables the creation of a more balanced dataset

- [PET Scans](https://www.sciencedirect.com/science/article/pii/S0168169921002969)
- [Macular Degeneration (ProGAN)](https://jamanetwork.com/journals/jamaophthalmology/article-abstract/2720489)
- [Brain Scan datasets](https://www.kaggle.com/datasets?search=brain+&tags=13207-Computer+Vision)

### Identifying Fake Images

[CIFAKE: Image Classification and Explainable Identification of AI-Generated Synthetic Images](https://arxiv.org/abs/2303.14126)
[cifake](https://www.kaggle.com/datasets/birdy654/cifake-real-and-ai-generated-synthetic-images)

### Monitoring Plant Health

[Afrocentric (African) Crop Dataset](https://www.kaggle.com/datasets/responsibleailab/crop-disease-ghana)

### Traffic Signs

[Traffic Sign Recognition repo](https://github.com/alexandrosstergiou/Traffic-Sign-Recognition-basd-on-Synthesised-Training-Data)

### Challenges

### Domain shift

> ### Circular use of synthetic data generation

Reviewer: Very interesting section imo, might go nicely with a graph showing accuracy degrading vs. number of times synthetic data has been "recycled" (if something like that exists) 😄
